### PR TITLE
Pulse Compiler `SetSequence` pass

### PR DESCRIPTION
**Pull request by TsafrirA**
### Summary
A `SetSequence` pass is introduced. The pass transforms Pulse's `SequenceIR` by setting the sequence of elements (adding edges to the graph). A second analysis pass is introduced - `MapMixedFrame` - which is needed for sequencing of parallel alignments.

### Details and comments
When `SequenceIR` object is populated, its `sequence` property includes only nodes and no edges. Setting the sequence (adding the edges to the graph) depends on the alignment type of the object. The pass `SetSequence` adds the edges to the graph according to the alignment type.

For parallel alignment, a mapping of `PulseTarget` and `Frame` to `MixedFrame`s is needed. For this purpose, an analysis pass `MapMixedFrame` is also added. The pass recurses through the IR, and identifies creates the mapping.

### CODE REVIEW

File: qiskit/pulse/transforms/alignments.py [hunk 1]
```diff
@@ -44,6 +46,22 @@ def align(self, schedule: Schedule) -> Schedule:
         """
         pass
 
+    @abc.abstractmethod
+    def set_sequence(self, sequence: PyDAG, **kwargs) -> None:
```
Comment (TsafrirA) [inline qiskit/pulse/transforms/alignments.py]: I am not sure kwargs is the way to go here. The difficulty is that parallel alignment needs the mixed frames mapping, but sequential alignment doesn't. It's also easier to just send the mapping to all alignment types, without checking the alignment type.

File: test/python/pulse/compiler_passes/test_set_sequence.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+
+    ddt_named_data = [["align_left", AlignLeft()], ["align_right", AlignRight()]]
+
+    def _get_pm(self) -> PulseIrTranspiler:
+        pm = PulseIrTranspiler()
+        pm.append(MapMixedFrame())
+        pm.append(SetSequence())
+        return pm
+
+    @named_data(*ddt_named_data)
```
Comment (TsafrirA) [inline test/python/pulse/compiler_passes/test_set_sequence.py:61]: I tested everything against all relevant alignment types. It's kind of an overkill, with the logic located at a parent abstract class.

File: qiskit/pulse/compiler/passes/map_mixed_frames.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        super().__init__(target=None)
+        self.mixed_frames_mapping = defaultdict(set)
+
+    def run(
+        self,
+        passmanager_ir: SequenceIR,
+    ) -> None:
+
+        self._analyze_mixed_frames_in_sequence(passmanager_ir)
+        self.property_set["mixed_frames_mapping"] = self.mixed_frames_mapping
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/map_mixed_frames.py]:
```suggestion
        mixed_frames_mapping = defaultdict(set)
        self._analyze_mixed_frames_in_sequence(passmanager_ir, mixed_frames_mapping)
        self.property_set["mixed_frames_mapping"] = mixed_frames_mapping
```
Can we implement the logic like this? Statefull pass sometimes causes edge cases. For example, in the current implementation some unexpected things could happen if you run the same pass manager instance twice for different inputs.

File: qiskit/pulse/compiler/passes/map_mixed_frames.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    def _analyze_mixed_frames_in_sequence(self, prog: SequenceIR) -> None:
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/map_mixed_frames.py]: Do you need this recursive call? If I understand correctly you can use `SequenceIR.inst_targets` to list all targets first and then create the map from the list. If we implement this recursive call on pass side, probably we can remove the method to make the IR simpler.

File: qiskit/pulse/compiler/passes/map_mixed_frames.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+from __future__ import annotations
+from collections import defaultdict
+
+from qiskit.pulse.compiler.basepasses import AnalysisPass
+from qiskit.pulse.ir import SequenceIR
+from qiskit.pulse.model import MixedFrame
+
+
+class MapMixedFrame(AnalysisPass):
+    """Map the dependencies of all ``MixedFrame``s on ``PulseTaraget`` and ``Frame``.
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/map_mixed_frames.py]:
```suggestion
    """Map the dependencies of all :class:`.MixedFrame`s on :class:`.PulseTaraget` and :class:`.Frame`.
```
Can you add cross ref?

File: qiskit/pulse/compiler/passes/map_mixed_frames.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    The pass recursively scans the ``SequenceIR``, identifies all ``MixedFrame``s and
+    tracks the dependencies of them on ``PulseTarget`` and ``Frame``. The analysis result
+    is added as a dictionary to the property set under key "mixed_frames_mapping". The
+    added dictionary is keyed on every ``PulseTarget`` and ``Frame`` in ``SequenceIR``
+    with the value being a set of all ``MixedFrame``s associated with the key.
+    """
+
+    def __init__(self):
+        """Create new MapMixedFrames pass"""
+        super().__init__(target=None)
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/map_mixed_frames.py:41]: This violates typehint (though type is not important in Python). Do you want to remove Target from the base passes?

File: qiskit/pulse/compiler/passes/map_mixed_frames.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            # Sub Block
+            if isinstance(elm, SequenceIR):
+                self._analyze_mixed_frames_in_sequence(elm)
+            # Pulse Instruction
+            else:
+                if isinstance(inst_target := elm.inst_target, MixedFrame):
+                    self.mixed_frames_mapping[inst_target.frame].add(inst_target)
+                    self.mixed_frames_mapping[inst_target.pulse_target].add(inst_target)
+
+    def __hash__(self):
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/map_mixed_frames.py:56]: Please also define `__eq__`

File: qiskit/pulse/compiler/passes/set_sequence.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        self,
+        passmanager_ir: SequenceIR,
+    ) -> SequenceIR:
+
+        self._set_sequence_recursion(passmanager_ir)
+        return passmanager_ir
+
+    def _set_sequence_recursion(self, prog: SequenceIR) -> None:
+        """A helper function to recurse through the sequence"""
+        prog.alignment.set_sequence(
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/set_sequence.py]: I don't think this is right direction. Alignment is just a type information and logic must be implemented on the pass side. We can just rely on the type information and ignore the alignment logic in the existing alignment kind instances. I would use [singledispatchmethod](https://docs.python.org/3/library/functools.html#functools.singledispatchmethod) to run type dependent sequencing. Of course this limits flexibility of adding new alignments, i.e. addition of a new alignment must always come with update of this pass, but I think this should be fine:
- One also needs to update builder command anyways
- So far I've never seen the request for new alignment (some people requested center alignment but no issue is written)

File: qiskit/pulse/transforms/alignments.py (hunk 1 above)
Comment (nkanazawa1989) [reply qiskit/pulse/transforms/alignments.py]: I think the logic must be moved to pass itself with dispatcher. Then every pass can access the mapping via `self.property_set`.

File: qiskit/pulse/compiler/passes/map_mixed_frames.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        """A helper function to recurse through the sequence while mapping mixed frame dependency"""
+        for elm in prog.elements():
+            # Sub Block
+            if isinstance(elm, SequenceIR):
+                self._analyze_mixed_frames_in_sequence(elm)
+            # Pulse Instruction
+            else:
+                if isinstance(inst_target := elm.inst_target, MixedFrame):
+                    self.mixed_frames_mapping[inst_target.frame].add(inst_target)
+                    self.mixed_frames_mapping[inst_target.pulse_target].add(inst_target)
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/map_mixed_frames.py]: Do you assume there is port only instruction? Maybe `SequenceIR.inst_targets` should only return `list[Frame|MixedFrame]`? i.e. virtual frame / frame with port

File: qiskit/pulse/transforms/alignments.py [hunk 10]
```diff
@@ -87,7 +105,110 @@ def __repr__(self):
...
+                # In, Out node
+                continue
+            node = sequence.get_node_data(ind)
+            node_mixed_frames = set()
+
+            # if isinstance(node, SequenceIR):
+            #     inst_targets = node.inst_targets
+            # else:
+            #     inst_targets = [node.inst_target]
+            if hasattr(node, "inst_targets"):
```
Comment (nkanazawa1989) [inline qiskit/pulse/transforms/alignments.py]: What is the reason you prefer the attribute name rather than the type? (and the reason you leave above code with comment)

File: test/python/pulse/compiler_passes/test_map_mixed_frames.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        self.assertEqual(len(mapping), 3)
+        self.assertEqual(mapping[mf.pulse_target], {mf, mf2})
+        self.assertEqual(mapping[mf.frame], {mf})
+
+    def test_with_several_inst_target_types(self):
+        """test with different inst_target types"""
+        ir_example = SequenceIR(AlignLeft())
+        mf = MixedFrame(Qubit(0), QubitFrame(1))
+        ir_example.append(Play(Constant(100, 0.1), mixed_frame=mf))
+        ir_example.append(Delay(100, target=Qubit(2)))
```
Comment (nkanazawa1989) [inline test/python/pulse/compiler_passes/test_map_mixed_frames.py:63]: Hmm maybe port only instruction make sense. This is very much compatible with `QuantumCircuit.delay` taking only qubits (so we can easily write transformation pass/scheduler)

File: test/python/pulse/compiler_passes/test_set_sequence.py [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+from .utils import PulseIrTranspiler
+
+
+@ddt
+class TestSetSequenceParallelAlignment(QiskitTestCase):
+    """Test SetSequence pass with Parallel Alignment"""
+
+    ddt_named_data = [["align_left", AlignLeft()], ["align_right", AlignRight()]]
+
+    def _get_pm(self) -> PulseIrTranspiler:
```
Comment (nkanazawa1989) [inline test/python/pulse/compiler_passes/test_set_sequence.py]: This could be moved to `SetUp` method (and this is why set up method exists)

File: test/python/pulse/compiler_passes/test_set_sequence.py [hunk 13]
```diff
@@ -0,0 +1,379 @@
...
+        pm = PulseIrTranspiler()
+        pm.append(MapMixedFrame())
+        pm.append(SetSequence())
```
Comment (nkanazawa1989) [inline test/python/pulse/compiler_passes/test_set_sequence.py]:
```suggestion
        pm = PulseIrTranspiler([MapMixedFrame(), SetSequence()])
```

File: test/python/pulse/compiler_passes/test_set_sequence.py [hunk 14]
```diff
@@ -0,0 +1,379 @@
...
+    @unpack
+    def test_no_mapping_pass_error(self, alignment):
+        """test that running without MapMixedFrame pass raises a PulseError"""
+
+        pm = PulseIrTranspiler()
+        pm.append(SetSequence())
+        ir_example = SequenceIR(alignment)
+        ir_example.append(Play(Constant(100, 0.1), mixed_frame=MixedFrame(Qubit(0), QubitFrame(1))))
+
+        with self.assertRaises(PulseError):
```
Comment (nkanazawa1989) [inline test/python/pulse/compiler_passes/test_set_sequence.py]: Maybe it's good timing to add new error type for compiler. More categorized errors allow us to write more trustable error handling logic, i.e. if we want to ignore a particular error we can just catch that error type. If we only have PulseError this could accidentally catch unexpected bugs.

File: test/python/pulse/compiler_passes/test_set_sequence.py [hunk 15]
```diff
@@ -0,0 +1,379 @@
...
+    @unpack
+    def test_single_instruction(self, alignment):
+        """test with a single instruction"""
+
+        ir_example = SequenceIR(alignment)
+        ir_example.append(Play(Constant(100, 0.1), mixed_frame=MixedFrame(Qubit(0), QubitFrame(1))))
+
+        ir_example = self._get_pm().run(ir_example)
+        edge_list = ir_example.sequence.edge_list()
+        self.assertEqual(len(edge_list), 2)
```
Comment (nkanazawa1989) [inline test/python/pulse/compiler_passes/test_set_sequence.py]: Maybe this is redundant.

File: test/python/pulse/compiler_passes/test_set_sequence.py [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+        self.assertTrue((0, 2) in edge_list)
+        self.assertTrue((2, 1) in edge_list)
+
+    # TODO: Take care of this weird edge case
+    # def test_instruction_not_in_mapping(self):
+    #     """test with an instruction which is not in the mapping"""
+    #
+    #     ir_example = SequenceIR(AlignLeft())
+    #     ir_example.append(Play(Constant(100, 0.1), mixed_frame=MixedFrame(Qubit(0), QubitFrame(1))))
+    #     ir_example.append(Delay(100, target=Qubit(5)))
```
Comment (nkanazawa1989) [inline test/python/pulse/compiler_passes/test_set_sequence.py]: Oh this is an interesting edge case. Probably this can be addressed with a resource mapping data. Let's leave this test, but not with comments. Please use https://docs.python.org/3/library/unittest.html#skipping-tests-and-expected-failures

File: test/python/pulse/compiler_passes/test_set_sequence.py [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+        self.assertEqual(len(edge_list), 5)
+        self.assertTrue((0, 2) in edge_list)
+        self.assertTrue((2, 3) in edge_list)
+        self.assertTrue((2, 4) in edge_list)
+        self.assertTrue((3, 1) in edge_list)
+        self.assertTrue((4, 1) in edge_list)
+
+    @named_data(*ddt_named_data)
+    @unpack
+    def test_frame_instruction_broadcasting_to_children(self, alignment):
```
Comment (nkanazawa1989) [inline test/python/pulse/compiler_passes/test_set_sequence.py]: Is the children right word? I see this operation is the broadcast to all mixed frame.

File: test/python/pulse/compiler_passes/test_set_sequence.py [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+        """test with an instruction which is defined on a Frame and is broadcasted to several children"""
+
+        ir_example = SequenceIR(alignment)
+        ir_example.append(ShiftPhase(100, frame=QubitFrame(0)))
+        ir_example.append(Play(Constant(100, 0.1), mixed_frame=MixedFrame(Qubit(0), QubitFrame(0))))
+        ir_example.append(Play(Constant(100, 0.1), mixed_frame=MixedFrame(Qubit(1), QubitFrame(0))))
```
Comment (nkanazawa1989) [inline test/python/pulse/compiler_passes/test_set_sequence.py:155]: Can you add one more instruction that doesn't take QubitFrame(0)? e.g. `MixedFrame(Qubit(1), QubitFrame(1))`

File: test/python/pulse/compiler_passes/test_set_sequence.py [hunk 19]
```diff
@@ -0,0 +1,379 @@
...
+    AlignFunc,
+    AlignEquispaced,
+)
+from qiskit.pulse.compiler import MapMixedFrame, SetSequence
+from qiskit.pulse.exceptions import PulseError
+from .utils import PulseIrTranspiler
+
+
+@ddt
+class TestSetSequenceParallelAlignment(QiskitTestCase):
```
Comment (nkanazawa1989) [inline test/python/pulse/compiler_passes/test_set_sequence.py:52]: In a typical circuit pass test, we use a public conversion function and test program is provided in a form of circuit, rather than a DAG circuit. Do you want to take the same pattern? I'm also fine with current pattern.
https://github.com/Qiskit/qiskit/blob/8b6e4fe0fe55fd96e182e526860321db702a7a54/test/python/transpiler/test_apply_layout.py#L36-L49

File: test/python/pulse/compiler_passes/utils.py
Comment (nkanazawa1989) [inline test/python/pulse/compiler_passes/utils.py:1]: Why do you need this file?

**Review by nkanazawa1989 — CHANGES_REQUESTED** [review]
Thanks @TsafrirA . The sequence logic looks good to me. My main concern is about implementation of sequence pass, and I think the sequence logic must be implemented inside the pass, rather than delegating it to the alignment object.

Here is another question. When we talk to a hardware, I think they request mixed frames. However, in the current implementation, broadcast is not really implemented. Instead, the sequence pass just create edges to a common node to broadcast. Do you want to write a separate pass for broadcast? We can also do this broadcast inside the sequencing pass with slight modification.

File: test/python/pulse/compiler_passes/utils.py
Comment (TsafrirA) [reply test/python/pulse/compiler_passes/utils.py:1]: I wanted an IR->IR compiler for the tests. I stashed it here.

File: test/python/pulse/compiler_passes/test_set_sequence.py (hunk 19 above)
Comment (TsafrirA) [reply test/python/pulse/compiler_passes/test_set_sequence.py:52]: I had the tests in terms of IR, so I left it like this. Once the builder is set up for the new model, I can modify the tests to be based on ScheduleBlock.

File: test/python/pulse/compiler_passes/test_set_sequence.py (hunk 17 above)
Comment (TsafrirA) [reply test/python/pulse/compiler_passes/test_set_sequence.py]: Yeah, it's probably a bad terminology on my part. I meant dependent instructions.

File: qiskit/pulse/compiler/passes/map_mixed_frames.py (hunk 6 above)
Comment (TsafrirA) [reply qiskit/pulse/compiler/passes/map_mixed_frames.py:41]: As I commented on the original compiler PR, we don't always need Target. I think we can make it optional.

File: qiskit/pulse/compiler/passes/map_mixed_frames.py (hunk 3 above)
Comment (TsafrirA) [reply qiskit/pulse/compiler/passes/map_mixed_frames.py]: The issue of re-running a pass is something that bothered me too. Do we override existing data? Raise an error? There's a "global" decision to be made here, because it's applicable to all passes.

File: test/python/pulse/compiler_passes/utils.py
Comment (TsafrirA) [reply test/python/pulse/compiler_passes/utils.py:1]: This will not be needed once\if we move on the ScheduleBlock based tests, instead of IR based.

File: qiskit/pulse/compiler/passes/map_mixed_frames.py (hunk 9 above)
Comment (TsafrirA) [reply qiskit/pulse/compiler/passes/map_mixed_frames.py]: Delay on a `PulseTarget` is the case I was considering here.

File: qiskit/pulse/transforms/alignments.py (hunk 10 above)
Comment (TsafrirA) [reply qiskit/pulse/transforms/alignments.py]: The cyclic imports issue. I couldn't import `SequenceIR` there.

File: test/python/pulse/compiler_passes/test_set_sequence.py (hunk 14 above)
Comment (TsafrirA) [reply test/python/pulse/compiler_passes/test_set_sequence.py]: Sounds good!

File: qiskit/pulse/compiler/passes/map_mixed_frames.py (hunk 4 above)
Comment (TsafrirA) [reply qiskit/pulse/compiler/passes/map_mixed_frames.py]: Nice catch. Thanks.

File: qiskit/pulse/compiler/passes/map_mixed_frames.py (hunk 3 above)
Comment (TsafrirA) [reply qiskit/pulse/compiler/passes/map_mixed_frames.py]: For the current pass, I modified it per your suggestion.

File: qiskit/pulse/compiler/passes/set_sequence.py (hunk 8 above)
Comment (TsafrirA) [reply qiskit/pulse/compiler/passes/set_sequence.py]: The scenario I had in mind was indeed addition of new alignments. But if this not something we are worried about (and you make good points), I'll change the logic.

File: qiskit/pulse/transforms/alignments.py (hunk 10 above)
Comment (TsafrirA) [reply qiskit/pulse/transforms/alignments.py]: (This stops being relevant as we move this to the pass itself)

File: test/python/pulse/compiler_passes/test_set_sequence.py (hunk 15 above)
Comment (TsafrirA) [reply test/python/pulse/compiler_passes/test_set_sequence.py]: I'll remove it.

File: test/python/pulse/compiler_passes/test_set_sequence.py (hunk 12 above)
Comment (TsafrirA) [reply test/python/pulse/compiler_passes/test_set_sequence.py]: Out of curiosity, what's the benefit of using `setUp`?

File: qiskit/pulse/compiler/passes/map_mixed_frames.py (hunk 3 above)
Comment (nkanazawa1989) [reply qiskit/pulse/compiler/passes/map_mixed_frames.py]: There is definitely a demand for rerunning the pass manager, for example https://github.com/Qiskit/qiskit/issues/11784. I prefer pass that doesn't have state, except for the user provided arguments.

File: qiskit/pulse/compiler/passes/set_sequence.py [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+        Nested :class:`.SequenceIR` objects are sequenced recursively.
+
+        Args:
+            alignment: The IR alignment.
+            sequence: The graph object to be sequenced.
+
+        Raises:
+            PulseCompilerError: if ``property_set`` does not include a mixed_frames_mapping dictionary.
+        """
+        if self.property_set["mixed_frames_mapping"] is None:
```
Comment (nkanazawa1989) [inline qiskit/pulse/compiler/passes/set_sequence.py]: Input type dependent check would be annoying for users. I would move to the `run` method and say the pass is dependent of `MapMixedFrame` in the `.. notes::` in the class doc (that's fair because sequential alignment context may contain internal parallel alignment context).

File: test/python/pulse/compiler_passes/test_set_sequence.py (hunk 12 above)
Comment (nkanazawa1989) [reply test/python/pulse/compiler_passes/test_set_sequence.py]: It's called in advance to call each `test_` function. If there is a bug in pass manager construction it's treated as an exception rather than the test failure. Also you can make the test code a bit simpler, i.e. `self.pm.run(...)` vs `self._get_pm().run(...)`.

File: test/python/pulse/compiler_passes/utils.py
Comment (nkanazawa1989) [reply test/python/pulse/compiler_passes/utils.py:1]: sounds good.

**Review by nkanazawa1989 — COMMENTED** [review]
Thanks, it looks almost good to go. Just a few minor comments before approval.

File: qiskit/pulse/compiler/passes/map_mixed_frames.py (hunk 3 above)
Comment (TsafrirA) [reply qiskit/pulse/compiler/passes/map_mixed_frames.py]: In that case, should we remove all edges from the sequence before sequencing?

File: test/python/pulse/compiler_passes/test_set_sequence.py (hunk 12 above)
Comment (TsafrirA) [reply test/python/pulse/compiler_passes/test_set_sequence.py]: Thanks!

**Review by nkanazawa1989 — APPROVED** [review]
LGTM thanks @TsafrirA
